File: Algorithms/109_ConvertSortedListToBinarySearchTree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a list node
    # @return a tree node
    def sortedListToBST(self, head):
        n = 0
        p = head
        while p:
            n += 1
            p = p.next
        self.listNode = head
        return self.sortedSubListToBST(0, n-1)

    def sortedSubListToBST(self, start, end):
        if start > end: return None
        mid = start + (end - start) / 2
        left = self.sortedSubListToBST(start, mid-1)
        parent = TreeNode(self.listNode.val)
        self.listNode = self.listNode.next
        parent.left = left
        parent.right = self.sortedSubListToBST(mid+1, end)
        return parent

File: Algorithms/109_ConvertSortedListToBinarySearchTree.cpp
#include "leetcode.h"
class Solution {
public:
    TreeNode *sortedListToBST(ListNode *head) {
        int length = 0;
        ListNode* p = head;
        while (p != NULL) {
            length ++;
            p = p->next;
        }
        return sortedListToBST(head, 0, length-1);
    }

    TreeNode* sortedListToBST(ListNode* &list, int start, int end) {
        if (start > end) return NULL;
        int mid = start + (end - start) / 2;
        TreeNode *leftChild = sortedListToBST(list, start, mid-1);
        TreeNode *parent = new TreeNode(list->val);
        parent->left = leftChild;
        list = list->next;
        parent->right = sortedListToBST(list, mid+1, end);
        return parent;
    } 
    /*
    TreeNode *sortedListToBST(ListNode *head) {
        if (head == NULL) {
            return NULL;
        }
        if (head->next == NULL) {
            return new TreeNode(head->val);
        }
        ListNode* slow = head;
        ListNode* fast = head->next->next;
        while (fast != NULL && fast->next != NULL) {
            fast = fast->next->next;
            slow = slow->next;
        }
        TreeNode* root = new TreeNode(slow->next->val);
        root->right = sortedListToBST(slow->next->next);
        slow->next = NULL;
        root->left = sortedListToBST(head);
        return root;
    }
    */
};
int main() {
    Solution s;
    ListNode* head = new ListNode(1);
    s.sortedListToBST(head);

    return 0;
}
